spectrum-network/tests/integration_tests/mod.rs

```diff
@@ -0,0 +1,661 @@
+
+    // Collect messages from the peers. Note that the while loop below will end since all tasks that
+    // use clones of `msg_tx` and `fake_msg_tx` are guaranteed to drop, leading to the senders
+    // dropping too.
+    let mut nc_peer_0 = vec![];
+    let mut nc_peer_1 = vec![];
+    let mut prot_peer_0: Vec<SyncMessage> = vec![];
+    let mut prot_peer_1: Vec<FakeSyncMessage> = vec![];
+    while let Some(c) = combined_stream.next().await {
+        match c {
+            C::SyncMsg((peer, msg)) => match msg {
+                Msg::NetworkController(nc_msg) => match peer {
+                    Peer::First => nc_peer_0.push(nc_msg),
+                    Peer::Second => nc_peer_1.push(nc_msg),
+                },
+                Msg::Protocol(p_msg) => prot_peer_0.push(p_msg),
+            },
+            C::FakeMsg((peer, msg)) => match msg {
+                Msg::NetworkController(nc_msg) => match peer {
+                    Peer::First => nc_peer_0.push(nc_msg),
+                    Peer::Second => nc_peer_1.push(nc_msg),
+                },
+                Msg::Protocol(p_msg) => prot_peer_1.push(p_msg),
+            },
+        }
+    }
+
+    dbg!(&nc_peer_0);
+    dbg!(&nc_peer_1);
+    dbg!(&prot_peer_0);
+    dbg!(&prot_peer_1);
+
+    let protocol_id = ProtocolId::from(0);
+    let protocol_ver = ProtocolVer::from(1);
+    let expected_nc_peer_0 = vec![
+        NetworkControllerOut::ConnectedWithInboundPeer(local_peer_id_1),
+        NetworkControllerOut::ProtocolPendingApprove {
+            peer_id: local_peer_id_1,
+            protocol_id,
+        },
+        NetworkControllerOut::ProtocolEnabled {
+            peer_id: local_peer_id_1,
+            protocol_id,
+            protocol_ver,
+        },
+        NetworkControllerOut::PeerPunished {
+            peer_id: local_peer_id_1,
+            reason: ReputationChange::MalformedMessage(MalformedMessage::UnknownFormat),
+        },
+        NetworkControllerOut::Disconnected {
+            peer_id: local_peer_id_1,
+            reason: ConnectionLossReason::Reset(ConnHandlerError::UnacceptablePeer),
+        },
+    ];
+
+    assert_eq!(expected_nc_peer_0, nc_peer_0);
+
+    let expected_nc_peer_1 = vec![
+        NetworkControllerOut::ConnectedWithOutboundPeer(local_peer_id_0),
+        NetworkControllerOut::ProtocolPendingEnable {
+            peer_id: local_peer_id_0,
+            protocol_id,
+        },
+        NetworkControllerOut::ProtocolEnabled {
+            peer_id: local_peer_id_0,
+            protocol_id,
+            protocol_ver,
+        },
+    ];
+    assert_eq!(expected_nc_peer_1, nc_peer_1);
+
+    assert_eq!(
+        prot_peer_0,
+        vec![SyncMessage::SyncMessageV1(SyncMessageV1::GetPeers),]
+    );
+    assert_eq!(
+        prot_peer_1,
+        vec![FakeSyncMessage::SyncMessageV1(FakeSyncMessageV1::FakeMsg),]
+    );
+}
+
+#[async_std::test]
+#[cfg_attr(not(feature = "test_peer_punish_too_slow"), ignore)]
```

Comment:
To run this integration test type: `cargo test integration_test_peer_punish_too_slow --features test_peer_punish_too_slow`